File: Java-V2/FindOptimalTransport.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import java.util.Stack;
import java.io.*;
import java.lang.Math;

public class FindOptimalTransport {

    private String[][] input;
    private String[][] minimumcell;
    private String[][] steppingstone;
    private Route[] routes;
    private Warehouse[] warehouses;
    private Factory[] factories;

    /**
     * 
     * @param filename
     */
    private int[] readIn(String filename) {
        int[] values;
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            BufferedReader br = new BufferedReader(new FileReader(filename));
            // get # of lines
            int rows = 0;
            int columns = 0;
            int counter = 0;
            String line;
            while ((line = reader.readLine()) != null) {
                if (counter == 0) {
                    String[] line1 = line.split(" ");
                    columns = line1.length;
                }
                counter++;
                rows++;
            }
            reader.close();

            input = new String[rows][columns];
            int countrow = 0;
            while ((line = br.readLine()) != null) {
                input[countrow] = line.split(" ");
                countrow++;
            }
            br.close();

            factories = new Factory[rows - 2];
            for (int i = 0; i < rows - 1 - 1; i++) {
                factories[i] = new Factory(Integer.parseInt(input[i + 1][columns - 1].replaceAll("\\s", "")), i + 1);
            }

            warehouses = new Warehouse[columns - 2];
            for (int i = 0; i < columns - 1 - 1; i++) {
                warehouses[i] = new Warehouse(Integer.parseInt(input[rows - 1][i + 1].replaceAll("\\s", "")), i + 1);
            }

            routes = new Route[(rows - 2) * (columns - 2)];
            int index = 0;
            for (int i = 0; i < factories.length; i++) {
                for (int j = 0; j < warehouses.length; j++) {
                    routes[index] = new Route(factories[i], warehouses[j],
                            Integer.parseInt(input[i + 1][j + 1].replaceAll("\\s", "")), 0);
                    index++;
                }
            }

            minimumcell = new String[rows][columns];
            steppingstone = new String[rows][columns];
            for (int i = 0; i < minimumcell[0].length; i++) {
                minimumcell[0][i] = input[0][i];
            }
            minimumcell[0][0] = "INITIAL";
            for (int i = 0; i < minimumcell[0].length; i++) {
                steppingstone[0][i] = input[0][i];
            }
            steppingstone[0][0] = "FINAL";
            minimumcell[rows - 1] = input[rows - 1];
            steppingstone[rows - 1] = input[rows - 1];

            for (int i = 1; i < rows - 1; i++) {
                minimumcell[i][0] = input[i][0];
                steppingstone[i][0] = input[i][0];
                minimumcell[i][columns - 1] = input[i][columns - 1];
                steppingstone[i][columns - 1] = input[i][columns - 1];

            }

            for (int i = 1; i < rows - 1; i++) {
                for (int j = 1; j < columns - 1; j++) {
                    minimumcell[i][j] = "0";
                    steppingstone[i][j] = "0";
                }
            }
            values = new int[] { rows, columns };

        }

        catch (FileNotFoundException ex) {
            System.out.println("File '" + filename + "' not found'");
            values = new int[] { 0, 0 };
        } catch (IOException ex) {
            System.out.println("Error reading file '" + filename + "'");
            values = new int[] { 0, 0 };
        }
        return values;

    }

    private void transferMinimumCell(Route route, int amount) {
        route.setAmount(route.getAmount() + amount);
        factories[route.getFactory().getId() - 1]
                .setDelivered(factories[route.getFactory().getId() - 1].getDelivered() + amount);
        warehouses[route.getWarehouse().getId() - 1]
                .setReceived(warehouses[route.getWarehouse().getId() - 1].getReceived() + amount);
        if (minimumcell[route.getFactory().getId()][route.getWarehouse().getId()] == "-") {
            minimumcell[route.getFactory().getId()][route.getWarehouse().getId()] = amount + "";
            steppingstone[route.getFactory().getId()][route.getWarehouse().getId()] = amount + "";
        } else {
            minimumcell[route.getFactory().getId()][route.getWarehouse().getId()] = (Integer.parseInt(
                    minimumcell[route.getFactory().getId()][route.getWarehouse().getId()].replaceAll("\\s", ""))
                    + amount) + "";
            steppingstone[route.getFactory().getId()][route.getWarehouse().getId()] = (Integer.parseInt(
                steppingstone[route.getFactory().getId()][route.getWarehouse().getId()].replaceAll("\\s", ""))
                    + amount) + "";
        }
    }

    private void transferSteppingStoneCell(Route route, int amount) {
        route.setAmount(route.getAmount() + amount);
        factories[route.getFactory().getId() - 1]
                .setDelivered(factories[route.getFactory().getId() - 1].getDelivered() + amount);
        warehouses[route.getWarehouse().getId() - 1]
                .setReceived(warehouses[route.getWarehouse().getId() - 1].getReceived() + amount);
        if (steppingstone[route.getFactory().getId()][route.getWarehouse().getId()] == "-") {
            steppingstone[route.getFactory().getId()][route.getWarehouse().getId()] = amount + "";
        } else {
            steppingstone[route.getFactory().getId()][route.getWarehouse().getId()] = (Integer.parseInt(
                steppingstone[route.getFactory().getId()][route.getWarehouse().getId()].replaceAll("\\s", ""))
                    + amount) + "";
        }
    }

    private Route getMinCost() {
        int cost = 0;
        int route_id = -1;
        boolean first = true;

        for (int i = 0; i < routes.length; i++) {
            if (first && !(routes[i].getWarehouse().getRequested() == routes[i].getWarehouse().getReceived()
                    || routes[i].getFactory().getProduced() == routes[i].getFactory().getDelivered())) {
                cost = routes[i].getCost();
                route_id = i;
                first = false;
            } else if (routes[i].getCost() < cost
                    && !(routes[i].getWarehouse().getRequested() == routes[i].getWarehouse().getReceived()
                            || routes[i].getFactory().getProduced() == routes[i].getFactory().getDelivered())) {
                cost = routes[i].getCost();
                route_id = i;
            }
        }

        return routes[route_id];
    }

    private Path copyAndAdd(Path path, Route route){
        Path newPath = new Path();
        for(int i = 0; i < path.getRoutes().size(); i++){
            newPath.getRoutes().add(path.getRoutes().get(i));
        }
        newPath.getRoutes().add(route);
        return newPath;
    }
    
    private ArrayList<Path> copyList(ArrayList<Path> pathlist){
    	ArrayList<Path> newlist = new ArrayList<Path>();
        for(int i = 0; i < pathlist.size(); i++){
            newlist.add(pathlist.get(i));
        }
        return newlist;
    }

    private boolean pathComplete(Path path){
        return path.getRoutes().get(0).getWarehouse().getId()==path.getRoutes().get(path.getRoutes().size()-1).getWarehouse().getId() &&
            path.getRoutes().size()>3 && path.getRoutes().size()%2 == 0;
    }


    private ArrayList<Path> getOptimalSequence() {
        ArrayList<Path> possiblepaths = new ArrayList<Path>();

        for (int i = 0; i < routes.length; i++) {
            if (routes[i].getAmount() == 0) {
                Path temp = new Path();
                temp.getRoutes().add(routes[i]);
                possiblepaths.add(temp);
            }
        }

        boolean complete = false;
        while(!complete){
            ArrayList<Path> newpaths = new ArrayList<Path>();
            ArrayList<Path> copy = copyList(possiblepaths);
            for(int i = 0; i < copy.size(); i++){
                Path temppath = copy.get(i);
                Route temproute = temppath.getRoutes().get(temppath.getRoutes().size()-1);
                ArrayList<Route> nextroutes = new ArrayList<Route>();
                if (temppath.getRoutes().size()%2==0) {
                    for (int j = 0; j < routes.length; j++) {
                        if ((routes[j].getAmount() != 0 && temppath.getRoutes().indexOf(routes[j]) == -1)
                                && routes[j].getWarehouse().getId() == temproute.getWarehouse().getId()
                                && routes[j] != temproute) {
                            nextroutes.add(routes[j]);
                        }
                    }
                } else {
                    for (int j = 0; j < routes.length; j++) {
                        if ((routes[j].getAmount() != 0 && temppath.getRoutes().indexOf(routes[j]) == -1)
                                && routes[j].getFactory().getId() == temproute.getFactory().getId()
                                && routes[j] != temproute) {
                                nextroutes.add(routes[j]);

                        }
                    }
                }
                if(nextroutes.size()==0 && !pathComplete(temppath)){
                }
                else if(pathComplete(temppath)&& nextroutes.size()!=0) {
                	newpaths.add(temppath);
                }
                else if(nextroutes.size()==0 && pathComplete(temppath)) {
                	newpaths.add(temppath);
                	
                }
                else{
                    for(int j = 0; j < nextroutes.size(); j++){
                        Path temp = copyAndAdd(temppath, nextroutes.get(j));
                        newpaths.add(temp);
                    }
                }
            }
            possiblepaths = newpaths;
            complete = true;
            for(int i = 0; i<possiblepaths.size(); i++){
                if(!pathComplete(possiblepaths.get(i))){
                    complete = false;
                }
            }
        }

        return possiblepaths;
    }



    private void minCellMethod(int supply) {
        boolean satisfied = false;
        while (!satisfied) {

            Route route = getMinCost();
            transferMinimumCell(route,
                    Math.min(
                            warehouses[route.getWarehouse().getId() - 1].getRequested()
                                    - warehouses[route.getWarehouse().getId() - 1].getReceived(),
                            factories[route.getFactory().getId() - 1].getProduced()
                                    - factories[route.getFactory().getId() - 1].getDelivered()));

            int delivered = 0;
            for (int i = 0; i < factories.length; i++) {
                delivered += factories[i].getDelivered();
            }
            if (delivered == supply) {
                satisfied = true;
            }
        }
    }

    private void steppingStoneMethod() {
        ArrayList<Path> paths = getOptimalSequence();
        boolean temp=true;
        while (paths.size()>0) {
            Path bestpath = paths.get(0);
            int bestcost = 0;

            for(int i=0; i<paths.size(); i++){
                int sum = 0;
                for(int j=0; j<paths.get(i).getRoutes().size(); j++){
                    sum+= paths.get(i).getRoutes().get(j).getCost()*Math.pow(-1, j);
                }
                paths.get(i).setUnitcost(sum);
                if(sum<bestcost){
                    bestpath = paths.get(i);
                    bestcost = sum;
                }
            }
            if(bestcost == 0){
                break;
            }

            int maxamount = bestpath.getRoutes().get(1).getAmount();
            for(int i=0; i<bestpath.getRoutes().size(); i++){
                if(Math.pow(-1, i)<0 && bestpath.getRoutes().get(i).getAmount()<maxamount){
                    maxamount = bestpath.getRoutes().get(i).getAmount();

                }
            }
            for(int i=0; i<bestpath.getRoutes().size(); i++){
            transferSteppingStoneCell(bestpath.getRoutes().get(i), (int)(maxamount*Math.pow(-1, i)));
            Route x = bestpath.getRoutes().get(i);
            }
            int coststep = 0;
            for(int i=0; i<routes.length; i++){
                coststep+=routes[i].getCost()*routes[i].getAmount();
            }
            
    
            paths = getOptimalSequence(); 

        }
    }

    private void print2D(String[][] toPrint) {
        // Loop through all rows
        for (int i = 0; i < toPrint.length; i++) {
            // Loop through all elements of current row
            for (int j = 0; j < toPrint[i].length; j++)
                if (j == 0) {
                   System.out.printf("%-9s", toPrint[i][j] + " ");
                } else {
                   System.out.printf("%-5s", toPrint[i][j] + " ");
                }
                
           System.out.print("\n");
        }
        System.out.println("\n");
    }

    public void findOptimalTransport(String filename) throws Exception {
        

        int[] rowcol = readIn(filename);
        int sumdemand = 0;
        int sumsupply = 0;
        for (int i = 0; i < warehouses.length; i++) {
            sumdemand += warehouses[i].getRequested();
        }
        for (int i = 0; i < factories.length; i++) {
            sumsupply += factories[i].getProduced();
        }
        if (sumdemand != sumsupply) {
            throw new Exception("Unblanced supply and demand");
        }

        System.out.println("\nInput: ");
        print2D(input);

        minCellMethod(sumsupply);

        int costmin = 0;
        for(int i=0; i<routes.length; i++){
            costmin+=routes[i].getCost()*routes[i].getAmount();
        }
        System.out.println("Minimum Cell Cost Method ($"+costmin+"):");

        print2D(minimumcell);

        int occupied = 0;
        for (int i = 1; i < minimumcell.length - 1; i++) {
            for (int j = 1; j < minimumcell[0].length - 1; j++) {
                if (minimumcell[i][j] != "0") {
                    occupied++;
                }
            }
        }
        if (occupied != rowcol[0] - 2 + rowcol[1] - 2 - 1) {
            throw new Exception("Degenerate case");
        }

        steppingStoneMethod();
        int coststep = 0;
        for(int i=0; i<routes.length; i++){
            coststep+=routes[i].getCost()*routes[i].getAmount();
        }
        System.out.println("Stepping Stone Method ($"+coststep+"):");

        print2D(steppingstone);

    }
}

File: Comprehensive-GO/findoptimaltransport.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
	"sync"
)

var input [][]string         //input matrix
var initial [][]string       //initial solution matrix
var steppingstone [][]string //stepping stone solution matrix
var routes []*Route          //list of routes
var warehouses []*Warehouse  //list of warehouses
var factories []*Factory     //list of factories
var wg sync.WaitGroup        //wait group

/*
Calculate current cost
*/
func currentCost() int {
	cost := 0
	for i := 0; i < len(routes); i++ {
		cost += routes[i].cost * routes[i].amount
	}
	return cost
}

/*
Reads in the input and initial matricies, populates the stepping
stone matrix, and populates the routes, factories, and routes slices
*/
func readIn(inputfile, solutionfile string) (int, int) {
	input = make([][]string, 0)
	initial = make([][]string, 0)

	//read in the input file
	file, err := os.Open(inputfile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.Split(scanner.Text(), " ")
		for i := 0; i < len(line); i++ {
			line[i] = strings.TrimSpace(line[i])
		}
		input = append(input, line)
	}

	//read in the initial solution
	file2, err2 := os.Open(solutionfile)
	if err2 != nil {
		log.Fatal(err2)
	}
	defer file2.Close()

	scanner2 := bufio.NewScanner(file2)
	for scanner2.Scan() {
		initial = append(initial, strings.Split(scanner2.Text(), " "))
	}

	if err2 := scanner2.Err(); err2 != nil {
		log.Fatal(err2)
	}
	row, col := len(input), len(input[0])

	//populate the stepping stone matrix
	steppingstone = make([][]string, row)
	for i := 0; i < row; i++ {
		steppingstone[i] = make([]string, col)
		for j := 0; j < col; j++ {
			if !(i == row-1 && j == col-1) {
				steppingstone[i][j] = initial[i][j]
			}
		}
	}

	//populate the factories slice
	factories = make([]*Factory, 0)
	for i := 0; i < row-2; i++ {
		val, err := strconv.Atoi(strings.TrimSpace(input[i+1][col-1]))
		if err == nil {
			factories = append(factories, newFactory(val, i+1))
		} else {
			panic("Error in making factories")
		}
	}

	//populate the warehouses slice
	warehouses = make([]*Warehouse, 0)
	for i := 0; i < col-2; i++ {
		val, err := strconv.Atoi(strings.TrimSpace(input[row-1][i+1]))
		if err == nil {
			warehouses = append(warehouses, newWarehouse(val, i+1))
		} else {
			panic("Error in making warehouses")
		}
	}

	//populate the routes slice
	routes = make([]*Route, 0)
	for i := 0; i < len(factories); i++ {
		for j := 0; j < len(warehouses); j++ {
			value, err := strconv.Atoi(strings.TrimSpace(input[i+1][j+1]))
			if err == nil {
				temproute := newRoute(factories[i], warehouses[j], value)
				if initial[i+1][j+1] != "-" {
					value2, err := strconv.Atoi(strings.TrimSpace(initial[i+1][j+1]))
					if err == nil {
						transferCells(temproute, value2)
					} else {
						panic("Error in making routes")
					}

				}
				routes = append(routes, temproute)

			} else {
				panic("Error in making routes")
			}
		}
	}

	return row, col
}

/*
Transfers the amount to route
*/
func transferCells(route *Route, amount int) {

	route.amount = route.amount + amount
	factories[route.factory.id-1].delivered =
		factories[route.factory.id-1].delivered + amount
	warehouses[route.warehouse.id-1].received =
		warehouses[route.warehouse.id-1].received + amount
	if steppingstone[route.factory.id][route.warehouse.id] == "-" {
		steppingstone[route.factory.id][route.warehouse.id] = strconv.Itoa(amount)
	} else {
		val, err := strconv.Atoi(strings.TrimSpace(steppingstone[route.factory.id][route.warehouse.id]))
		if err == nil {
			steppingstone[route.factory.id][route.warehouse.id] = strconv.Itoa(val + amount)
		} else {
			panic("Error while transfering cells")
		}
	}
}

/*
Finds the path from an empty cell route and calculates the marginal cost,
the path is returned through the channel ch
*/
func marginalCost(route *Route, ch chan *Path) {
	defer wg.Done()
	possiblepaths := make([]*Path, 0)

	temp := newPath()
	temp.routes = append(temp.routes, route)
	possiblepaths = append(possiblepaths, temp)

	complete := false
	//keeps adding routes until all paths are a dead end(removed) or complete
	for !complete {
		newpaths := make([]*Path, 0)
		copy := copyList(possiblepaths)
		for i := 0; i < len(copy); i++ {
			temppath := copy[i]
			temproute := temppath.routes[len(temppath.routes)-1]
			nextroutes := make([]*Route, 0)
			//if even length search vertically
			if len(temppath.routes)%2 == 0 {
				for j := 0; j < len(routes); j++ {
					if routes[j].amount != 0 && !containsRoute(temppath.routes, routes[j]) && routes[j].warehouse.id == temproute.warehouse.id && routes[j] != temproute {
						nextroutes = append(nextroutes, routes[j])
					}
				}
				//if odd length search horizontally
			} else {
				for j := 0; j < len(routes); j++ {
					if routes[j].amount != 0 && !containsRoute(temppath.routes, routes[j]) && routes[j].factory.id == temproute.factory.id && routes[j] != temproute {
						nextroutes = append(nextroutes, routes[j])

					}
				}
			}
			//check if paths are a dead end and need to be removed or if they are complete
			if len(nextroutes) == 0 && !temppath.pathComplete() {
			} else if temppath.pathComplete() && len(nextroutes) != 0 {
				newpaths = append(newpaths, temppath)
			} else if len(nextroutes) == 0 && temppath.pathComplete() {
				newpaths = append(newpaths, temppath)

			} else {
				for j := 0; j < len(nextroutes); j++ {
					temp := copyAndAdd(temppath, nextroutes[j])
					newpaths = append(newpaths, temp)
				}
			}
		}
		possiblepaths = newpaths
		complete = true
		//check if done looking for paths
		for i := 0; i < len(possiblepaths); i++ {
			if !possiblepaths[i].pathComplete() {
				complete = false
			}
		}
	}
	possiblepaths[0].unitcost = possiblepaths[0].getUnitCost()
	ch <- possiblepaths[0]

}

/*
Returns the optimal path for the current stepping stone step
*/
func getOptimalSequence() (*Path, int) {
	possiblepaths := make([]*Path, 0)
	ch := make(chan *Path, 1)

	//run a go routine for each empty cell to find the path
	for i := 0; i < len(routes); i++ {
		if routes[i].amount == 0 {
			wg.Add(1)
			go marginalCost(routes[i], ch)
		}
	}

	go func() {
		wg.Wait()
		close(ch)
	}()

	//gather all the paths
	for val := range ch {
		possiblepaths = append(possiblepaths, val)
	}

	//find the best path
	bestpath := possiblepaths[0]
	bestcost := 0

	for i := 0; i < len(possiblepaths); i++ {
		if possiblepaths[i].unitcost < bestcost {
			bestpath = possiblepaths[i]
			bestcost = possiblepaths[i].unitcost
		}
	}

	return bestpath, bestcost
}

/*
Finds the optimal transport iteratively given an input and initial solution filenames
*/
func findOptimalTransport(inputfile, solutionfile string) {
	//read in the input and initial solution

	file, err := os.Create("solution.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Create a buffered writer from the file
	bufferedWriter := bufio.NewWriter(file)

	/*_, err = bufferedWriter.WriteString(
		"Buffered string\n",
	)
	if err != nil {
		panic(err)
	}*/

	row, col := readIn(inputfile, solutionfile)

	//calculate the  total transportation cost
	oldcost := currentCost()

	//check if degenerate case
	occupied := 0
	for i := 1; i < row-1; i++ {
		for j := 1; j < col-1; j++ {
			if initial[i][j] != "0" && initial[i][j] != "-" {
				occupied++
			}
		}
	}
	if occupied != row-2+col-2-1 {
		_, err = bufferedWriter.WriteString(
			"Degenerate Case\n",
		)
		if err != nil {
			panic(err)
		}
		panic("Degenerate Case")
	} else {

		//while there is a best path with negative cost, keep iterating
		bestpath, cost := getOptimalSequence()
		for cost < 0 {

			//find the amount to transfer
			maxamount := bestpath.routes[1].amount
			for i := 0; i < len(bestpath.routes); i++ {
				if math.Pow(-1, float64(i)) < 0 && bestpath.routes[i].amount < maxamount {
					maxamount = bestpath.routes[i].amount

				}
			}
			//for each route in the path transfer the calculated amount
			for i := 0; i < len(bestpath.routes); i++ {
				transferCells(bestpath.routes[i], int(maxamount*int(math.Pow(-1, float64(i)))))

			}

			bestpath, cost = getOptimalSequence()

		}
		//calculate the new total transportation cost
		coststep := currentCost()

		//Print out the input and initial solution
		fmt.Println("Input")
		print2D(input)

		_, err = fmt.Fprint(bufferedWriter,
			"Input\r\n",
		)
		if err != nil {
			panic(err)
		}

		print2DFile(input, bufferedWriter)

		fmt.Printf("Initial Solution ($%d):\r\n", oldcost)
		print2D(initial)

		_, err = fmt.Fprintf(bufferedWriter, "Initial Solution ($%d):\r\n", oldcost)
		if err != nil {
			panic(err)
		}

		print2DFile(initial, bufferedWriter)

		//Print out the stepping stone solution
		fmt.Printf("Stepping Stone Method ($%d):\r\n", coststep)
		print2D(steppingstone)

		_, err = fmt.Fprintf(bufferedWriter, "Stepping Stone Method ($%d):\r\n", coststep)
		if err != nil {
			panic(err)
		}

		print2DFile(steppingstone, bufferedWriter)
		bufferedWriter.Flush()
	}
}
